### PR TITLE
Faulty deserializers generated when record is splitted in subsequent schema

### PR DESCRIPTION
**Pull request by flowenol**
Hi,

Recently we ran into this issue while doing some refactoring on our production schemas. The aim of the refactoring was to split the already existing record into two separate records preserving the backwards compatibility via the use of type aliases.

Example: record A with aliases B and C in the next version is changed into records B and C, both of them have an alias to A.

This resulted in faulty deserializers being generated, which were unable to read the data for both reading scenarios with schemas interchanged.

The fix for this bug required some changes around method reference caching in FastDeserializerGenerator and in various places the writer schema had to be altered with reader schema.

15f4495 contains the test cases with faulty deserializers code-gen
48e6b22 contains the fix with proper deserializers code-gen

cheers,
Peter

### CODE REVIEW

**Comment by gaojieliu** [comment]
@flowenol 
Thanks for this improvement.
Could you split this code change into two commits as this one: https://github.com/linkedin/avro-util/pull/72/commits ?
One commit contains all the code gen logic and related tests, and the other one only contains the generated classes.
With this way, it will be easier for review to understand what has been changed.

Thanks

**Comment by wikkoz** [comment]
Hi @gaojieliu 

I have rearranged commits as you wanted.

**Comment by gaojieliu** [comment]
Hi @wikkoz 
Thank you, will review your RB soon.

**Review by gaojieliu — CHANGES_REQUESTED** [review]
Sorry for the late review.
The code change looks good overall, and I left a minor comment.

File: avro-fastserde/src/test/java/com/linkedin/avro/fastserde/FastSerdeTestsSupport.java [hunk 1]
```diff
@@ -96,7 +98,27 @@ public static Schema createRecord(String name, Schema.Field... fields) {
...
+    return fixNullDefaultForField(new Schema.Field(name, unionSchema, null, null, Schema.Field.Order.ASCENDING));
+  }
+
+  private static Schema.Field fixNullDefaultForField(Schema.Field field) {
+    try {
+      Class<?> nullNodeClass;
+      if (AvroCompatibilityHelper.getRuntimeAvroVersion().earlierThan(AvroVersion.AVRO_1_9)) {
+        nullNodeClass = Class.forName("org.codehaus.jackson.node.NullNode");
+      } else {
+        nullNodeClass = Class.forName("com.fasterxml.jackson.databind.node.NullNode");
```
[the remark this answers is not in the document]
Comment (wikkoz) [reply avro-fastserde/src/test/java/com/linkedin/avro/fastserde/FastSerdeTestsSupport.java:110]: Yes here is JIRA for it
https://issues.apache.org/jira/browse/AVRO-1126

File: avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java [hunk 2]
```diff
@@ -1080,41 +1083,46 @@ private boolean schemaNotRegistered(final Schema schema) {
...
-        getUniqueName("deserialize" + schema.getName()));
+        getUniqueName("deserialize" + effectiveSchema.getName()));
 
     method._throws(IOException.class);
     method.param(Object.class, VAR_NAME_FOR_REUSE);
     method.param(Decoder.class, DECODER);
 
-    (read ? deserializeMethodMap : skipMethodMap).put(AvroCompatibilityHelper.getSchemaFullName(schema), method);
+    (read ? deserializeMethodMap : skipMethodMap).put(AvroCompatibilityHelper.getSchemaFullName(writerSchema)
+        + writerSchema.hashCode() + (readerSchema != null ? readerSchema.hashCode() : ""), method);
```
[the remark this answers is not in the document]
Comment (wikkoz) [reply avro-fastserde/src/main/java/com/linkedin/avro/fastserde/FastDeserializerGenerator.java]: Done

**Review by gaojieliu — APPROVED** [review]

**Comment by gaojieliu** [comment]
@wikkoz 
Please try this release: 
https://bintray.com/linkedin/maven/avro-util/0.2.25

**Pull request merged**